kv_store/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+#![no_std]
+
+pub mod path;
+pub use path::Path;
+
+#[cfg(feature = "tokio")]
+pub mod async_query;
+#[cfg(feature = "tokio")]
+pub use async_query::{requester, Requester};
+
+extern crate alloc;
+use alloc::{
+    boxed::Box,
+    collections::{btree_map::Entry, BTreeMap},
+};
+use core::{clone::Clone, ops::Bound};
+use edfsm::{Change, Drain, Fsm, Init, Input, Terminating};
+use serde::{Deserialize, Serialize};
+
+/// The event type of an Fsm
+pub type Event<M> = <M as Fsm>::E;
+
+/// The command type of an Fsm
+pub type Command<M> = <M as Fsm>::C;
+
+/// The input type of an Fsm
+pub type In<M> = Input<<M as Fsm>::C, <M as Fsm>::E>;
+
+/// The output message type of an Fsm for the purpose of this module.
+pub type Out<M> = <<M as Fsm>::SE as Drain>::Item;
+
+/// The effector/effects type of an Fsm
+pub type Effect<M> = <M as Fsm>::SE;
+
+/// The state type of an Fsm
+pub type State<M> = <M as Fsm>::S;
+
+/// A query to the KV store.
+///
+/// Type parameter `V` is the value type of the KV store. (The key type is `Path`.)
+/// The parameter `E` is type for events that can update a value. In other words,
+/// the inner state machine at each Path receives events type `E` and manages state type `V`.
+///
+/// The functions of type `RespondOne` and `RespondMany` are passed the query result.
+/// Results contain borrowed values `&V` which can't be passed to channels or
+/// other data structures.  The respond function may clone these to pass them on,
+/// or the function may interpret or aggregate borrowed values in place.
+///
+pub enum Query<V, E> {
+    /// Get the value at the given path, or None.
+    Get(Path, RespondOne<V, ()>),
+
+    /// Get the entries whose path starts with the given path,
+    /// including the entry for the path itself.
+    GetTree(Path, RespondMany<V, ()>),
+
+    /// Get the entries in the given range
+    GetRange((Bound<Path>, Bound<Path>), RespondMany<V, ()>),
+
+    /// Get all the entries
+    GetAll(RespondMany<V, ()>),
+
+    /// Get the value at the given path or None and emit an event for that path.
+    Upsert(Path, RespondOne<V, E>),
+
+    /// Get all the entries and emit an event for a particular (usually new) path.
+    Insert(RespondMany<V, Keyed<E>>),
+}
```

Comment:
Is it useful that these types remain public given the `Requester`?

Reply:
Yes very useful.  Requester is feature gated for tokio.   kv_store is allocated but no_std.